spatialos/UnrealGDK
# UNR-4030 Add ability to specify SpatialMultiWorkerSettings class from command line (#2245)

Add the ability to specify "SpatialMultiWorkerSettings" class from the command line.
--------------------------------------------------------------------
specify it via `-OverrideMultiWorkerSettingsClass=SoftClassPath`
like:
for BP:`-OverrideMultiWorkerSettingsClass=/Game/LoadBalancing/BP_BenchmarkGymSettings.BP_BenchmarkGymSettings_C`
for C++:`-OverrideMultiWorkerSettingsClass=/Script/SpatialGDK.SpatialMultiWorkerSettings`

BTW:
This solution follows the existing pattern used for `bOverrideMultiWorker`
This function has the side-effect behavior that we apply a change in what might otherwise be a const function.
`bool USpatialStatics::IsSpatialMultiWorkerEnabled(const UObject* WorldContextObject)`

In order to ensure that there is no inconsistency when using the command line override both the
`bEnableMultiWorker` boolean flag and
`SpatialWorldSettings::MultiWorkerSettingsClass` values are rewritten.

Additionally with this override we may no longer need the OverrideMultiWorker option but removing this involves changes to the offloading nfr tests and was therefore deemed outside the scope of this ticket, see [UNR-4065](https://improbableio.atlassian.net/browse/UNR-4065).

Actually the side-effect behavior does not feel to me like a proper solution (the Is.. function should be const), but because this work is blocking additional work and it follows an existing pattern I chose this solution.

#### Release note
REQUIRED: Add a release note to the `##Unreleased` section of CHANGELOG.md. You can find guidance for writing useful release notes [here](../SpatialGDK/Extras/internal-documentation/how-to-write-good-release-notes.md). Documentation changes are exempt from this requirement.

#### Tests
How did you test these changes prior to submitting this pull request?
1) LaunchSpatial.bat
2) Change GDKTestGyms Configuration Properties / Debugging / Command Arguments as:
`"$(SolutionDir)GDKTestGyms.uproject" -skipcompile /Game/Maps/benchmarkgym -ReadFromCommandLine -TotalPlayers=3 -PlayerDensity=3 -TotalNPCs=12 -MaxRoundTrip=150 -MaxLateness=150 -server -log LOG=nfrserver.txt -workerType UnrealWorker -stdout -nowrite -unattended -nologtimes -nopause -noin -messaging -SaveToUserDir -NoVerifyGC -windowed -resX=400 -resY=300 -OverrideMultiWorkerSettingsClass=/Game/LoadBalancing/BP_BenchmarkGymSettings:BP_BenchmarkGymSettings_C`
or 
`"$(SolutionDir)GDKTestGyms.uproject" -skipcompile /Game/Maps/benchmarkgym -ReadFromCommandLine -TotalPlayers=3 -PlayerDensity=3 -TotalNPCs=12 -MaxRoundTrip=150 -MaxLateness=150 -server -log LOG=nfrserver.txt -workerType UnrealWorker -stdout -nowrite -unattended -nologtimes -nopause -noin -messaging -SaveToUserDir -NoVerifyGC -windowed -resX=400 -resY=300 -OverrideMultiWorkerSettingsClass=/Script/SpatialGDK.SpatialMultiWorkerSettings`
3) Add a breakpoint at Engine\Plugins\UnrealGDK\SpatialGDK\Source\SpatialGDK\Private\Utils\SpatialStatics.cpp line:113
4) Start Debugging
5) Make sure `WorldSettings->MultiWorkerSettingsClass` is valid.

What automated tests are included in this PR? 
No

STRONGLY SUGGESTED: How can this be verified by QA?
There should be an error if specify a run classpath for MultiWorkerSettingsClass.
like this:
```
Assertion failed: WorldSettings->MultiWorkerSettingsClass != nullptr [File:C:\io\4.25-SpatialOSUnrealGDK\Engine\Plugins\UnrealGDK\SpatialGDK\Source\SpatialGDK\Private\Utils\SpatialStatics.cpp] [Line: 116] 
/Game/LoadBalancing/BP_BenchmarkGymSettings:BP_BenchmarkGymSettings_Cxx is not a valid class
UE4Editor.exe has triggered a breakpoint.
```

#### Documentation
How is this documented (for example: release note, upgrade guide, feature page, in-code documentation)?
- Add ability to specify `USpatialMultiWorkerSettings` class from command line. Specifying a `SoftClassPath` via `-OverrideMultiWorkerSettingsClass=MultiWorkerSettingsClassName`.
